backend/index.js

```diff
@@ -193,8 +229,63 @@ app.get("/api/categories", async (req, res) => {
   }
 });
 
-// Process JSON file
-processJsonFile();
+//Interact route
+app.post("/interact", async (req, res) => {
+  const { businessId, liked, saved, viewed, reviewed, rated, userId } =
+    req.body;
+  try {
+    const interaction = await prisma.interaction.upsert({
+      where: { userId_businessId: { userId, businessId } },
+      update: { liked, saved, viewed, reviewed, rated },
+      create: { userId, businessId, liked, saved, viewed, reviewed, rated },
+    });
+    res.status(200).json(interaction);
+  } catch (error) {
+    console.error("Error creating/updating interaction:", error);
+    res.status(500).json({ error: "Failed to create/update interaction" });
+  }
+});
+
+app.get("/api/favorites/:id", async (req, res) => {
+  const { id } = req.params;
+  try {
+    const favorites = await prisma.interaction.findMany({
+      where: { userId: parseInt(id), liked: true },
+      include: { Business: true },
+    });
+
+    const uniqueBusinesses = Array.from(
+      new Set(favorites.map((interaction) => interaction.businessId))
+    ).map((businessId) => {
+      return favorites.find(
+        (interaction) => interaction.businessId === businessId
+      ).Business;
+    });
+
+    res.json(uniqueBusinesses);
+  } catch (error) {
+    console.error("Error fetching favorites:", error);
+    res.status(500).json({ error: "Failed to fetch favorites" });
+  } finally {
+    await prisma.$disconnect();
+  }
+});
+
+// Endpoint to trigger data preprocessing
+app.get("/preprocess-data", (req, res) => {
```

Comment:
Just as a heasd up: if this is an expensive operation, you probably wouldn't want to make this publicly accessible (or add rate limiting) since malicious users could abuse it.  (E.g. create a ton of OS processes via the `exec()`).  But I presume these sort of security things are not in the scope of the course.

Reply:
Agree, doesn't seem like this should be a public endpoint. Probably better to make a command in your package.json, if you need to run this frequently. eg. "npm run data-processor".